python/btcspv/validate_spv.py

```diff
@@ -0,0 +1,170 @@
+    for tx_out in t.tx_outs:
+        b.extend(tx_out)
+    return b
+
+
+# TODO: add documentation to all of these
+def extract_merkle_root_le(h: bytes) -> bytes:
+    '''Extracts the transaction merkle root from a header (little-endian)'''
+    return h[36:68]
+
+
+def extract_prev_block_le(h: bytes) -> bytes:
+    '''Extracts the previous block's hash from a header (little-endian)'''
+    return h[4:36]
+
+
+def extract_prev_block_be(h: bytes) -> bytes:
+    '''Extracts the previous block's hash from a header (big-endian)'''
+    return extract_prev_block_le(h)[::-1]
+
+
+def prove(
+    txid: bytes, merkle_root: bytes, intermediate_nodes: bytes, index: int
+) -> bool:
+    '''Validates a tx inclusion in the block'''
+    if txid == merkle_root and index == 0 and len(intermediate_nodes) == 0:
+        return True
+
+    proof = txid + intermediate_nodes + merkle_root
+    return verify_proof(proof, index)
+
+
+def validate_header(header: RelayHeader) -> bool:
+    '''
+    Verifies a bitcoin header
+    Args:
+        header (object): The header as an object
+        header.raw (bytes): The bitcoin header
+        header.hash (bytes): The hash of the header
+        header.hash_le (bytes): The LE hash of the header
+        header.height (int): The height
+        header.prevhash (bytes): The hash of the previous header
+        header.merkle_root The merkle root of the header
+        header.merkle_root_le The LE merkle root
+
+    Returns:
+        (bool): True if valid header, else False
+    '''
+    # Check that HashLE is the correct hash of the raw header
+    headerHash = rutils.hash256(header.raw)
+    if headerHash != header.hash_le:
+        # TODO: Throw error instead of returning false
+        # throw new Error('Hash LE is not the correct hash of the header')
+        return False
+
+    # Check that HashLE is the reverse of Hash
+    reversedHash = header.hash[::-1]
+    if reversedHash != header.hash_le:
+        # throw new Error('HashLE is not the LE version of Hash');
+        return False
+
+    # Check that the MerkleRootLE is the correct MerkleRoot for the header
+    extractedMerkleRootLE = extract_merkle_root_le(header.raw)
+    if extractedMerkleRootLE != header.merkle_root_le:
+        # throw new Error(
+        # 'MerkleRootLE is not the correct merkle root of the header')
+        return False
+
+    # Check that MerkleRootLE is the reverse of MerkleRoot
+    reversedMerkleRoot = header.merkle_root[::-1]
+    if reversedMerkleRoot != header.merkle_root_le:
+        # throw new Error('MerkleRootLE is not the LE version of MerkleRoot');
+        return False
+
+    # Check that PrevHash is the correct PrevHash for the header
+    extractedPrevHash = extract_prev_block_be(header.raw)
+    if extractedPrevHash != header.prevhash:
+        # throw new Error(
+        # 'Prev hash is not the correct previous hash of the header');
+        return False
+
+    return True
+
+
+def validate_spvproof(proof: object) -> bool:
```

Comment:
Type should be `SPVProof`

`from btcspv.types import RelayHeader, SPVProof`

And then args just need to specify one argument, rather than the whole object structure

Reply:
attributes will be accessed like a python dictionary `proof['attribute_name']`

Reply:
Fixed, changed documentation for Args, will push in a minute